### DISCUSSION TITLE
Publishing more often by iterating more frequently (a toolchain discussion)

### DISCUSSION THREAD

**Discussion by semioticrobotic** [theopensourceway/guidebook#237] [Ideas]
I'd like to initiate a discussion about ways we might release new versions of the guide more frequently by iterating more rapidly.

I was lucky enough to share some of these thoughts with @quaid, @bproffitt, and @shaunix at the All Things Open conference recently; now I want marshal my thoughts on the matter by writing something here.

At the moment, our previous release is [more than three years old](https://github.com/theopensourceway/guidebook/releases/tag/v2.0). And while we've pushed useful and important [changes to the repository since then](https://github.com/theopensourceway/guidebook/commits/main/), we haven't brought all these together in a [new version](https://github.com/theopensourceway/guidebook/discussions/234). I still use (reference, cite, add to, etc.) The Open Source Way handbook frequently, and I'd like to see the resource continue to grow and thrive with new versions and activity that inspires others to jump in and contribute, too.

My assessment of the situation (accuracy yours to determine!) is that our publishing process is a bit overwrought. We've got a nice, streamlined [editorial process](https://github.com/theopensourceway/guidebook/blob/main/PROCESS.adoc), but getting materials from "polished, ready to go!" to "live, now part of the guide!" seems to be a sticking point. I wonder if our AsciiDoc-based toolchain might be too complicated, getting in our way more than propelling us forward.

I say this because I've recently been experimenting (productively) with new Git-based, Markdown-to-website generators that have made publishing prose from GitHub repositories like ours incredibly straightforward. The state of the art in this space has advanced quite a bit since we initiated our project, and some of these tools are much more useful for projects like ours than perhaps they were when we got started. [GitBook](https://www.gitbook.com/) (a SaaS solution available at no cost of open source projects) and [Docusaurus](https://docusaurus.io/) (a self-hosted option) come to mind, as they're the two with which I've become most familiar.

Tools like these *really* streamline the work of publishing Markdown-based documents as websites, and I think adopting something like this could accelerate our own practice here. They take the "legwork" out of publishing updates to handbook-style resources, reflecting changes to a codebase almost immediately after those changes land in the project's main branch. I recognize that adopting a tool like this would require a bit of refactoring in our codebase (AsciiDoc and Markdown are similar but not identical!), but I want to suggest that this hill may be worth the climb—especially if it means releasing more than once every half-decade.

What do others think?

**Comment by semioticrobotic** [comment]
It strikes me, too, that in the event that adopting a new publishing toolchain seems undesirable/intimidating for the community, we might at least consider automating our current publication process (which is already script-based) with GitHub Actions to remove the need for extensive human intervention every time a new release is warranted.

**Comment by bproffitt** [comment]
I think this is a fantastic idea, and I am glad you have stepped up to initiate this. I will commit myself and other Red Hat resources to helping with a switch to the MD format, however we can be of assistance.

**Reply by semioticrobotic** [reply]
Thanks for this affirmation, @bproffitt. I'm away for another few days but perhaps when I'm back in the saddle we could schedule an informal call to discuss in more detail. 

**Update:** Meeting discussion floated.

**Comment by semioticrobotic** [comment]
Thanks for a productive meeting today, @bproffitt, @shaunix, and @quaid. It was great to see you all again.

At the meeting, we decided to move ahead with a plan to refactor the Open Source Way guidebook and port it to Markdown for use with GitBook. To that end, we outlined the following next steps:

- [x] Clean up `guidebook` repository by moving instructional files to `the-project` (see https://github.com/theopensourceway/guidebook/commit/7bdd423ffb1fcad2481825e38aca6cc68d3e8b9d and https://github.com/theopensourceway/the-project/commit/268972c62094ce21086d0d160b1a4c2f8e091639)
- [x] Create new working branch of `guidebook` project (see working branch [`2025-refactor`](https://github.com/theopensourceway/guidebook/tree/2025-refactor))
- [x] Create GitBook account and add maintainers
- [x] Connect GitBook organization to project repository and working branch
- [x] Begin refactoring the guidebook by converting files from AsciiDoc to Markdown
- [x] Seek [GitBook sponsorship](https://docs.gitbook.com/account-management/plans/apply-for-the-non-profit-open-source-plan) as a community-driven open source project

I volunteered to get us started so I will just start at the top of the list and get rolling.

**Comment by semioticrobotic** [comment]
Well, it took a little time, but I finally have a [working prototype](https://the-open-source-way.gitbook.io/guidebook) ready for review:

`https://the-open-source-way.gitbook.io/guidebook`

@theopensourceway/maintainers, please take a look and see what you think!

At the moment, the GitBook site is tracking a working branch of the GitHub project, [`2025-refactor`](https://github.com/theopensourceway/guidebook/tree/2025-refactor). If and when we decide to move forward following review, we can merge that branch, set the GitBook to track `main`, and update URLs as necessary.

I have plenty of thoughts/notes on the project in light of recently touching each chapter, but I'll hold those for a moment until we've all had a chance to review and comment on the work.

Happy Thanksgiving to those celebrating! 🦃

**Comment by semioticrobotic** [comment]
Oh, and one other thing.

The `main` branch features several stubs of chapters-in-progress. I've moved these to a temporary location ([the-project/editorial/chapter-drafts](https://github.com/theopensourceway/the-project/tree/main/editorial/chapter-drafts)) for safe keeping while we hack around with the refactor.

Ultimately, what I would like to do is open a new working branch for each of these chapters, so authors can work simultaneously on bringing them to life and we can merge into `main` (and publish) when each is finished. But I would prefer to do this after we've first merged any changes to the `main` branch, as it may eventually have a new folder structure.

I understand that this represents something of a departure from our previous way of working, but I will eventually be proposing changes to that process, too, now that we're entertaining use of a different publication platform.

**Comment by bproffitt** [comment]
Okay, I have been away on PTO (_some_ of us know how to do that, BB ;) ), and will review this this week.

**Reply by semioticrobotic** [reply]
**P**lenty of **T**ime for **O**pen Source. I know how to do that. Thanks for reviewing!

**Comment by bproffitt** [comment]
Got into this and I have to say this is pretty great. The format of the final product is solid, and I love the navigation and display elements. I found the source code in the 025-refactor branch and the md files look pretty clean and consistent. A couple of questions:

- Were there any conversion-related issues/cleanup concerns?
- Do we care that the "Powered by Gitbook" stamp is all over the site? Part of me is like "cool, support our fellow open source projects" and part of me is also "it feels like a NASCAR sticker."

**Reply by semioticrobotic** [reply]
So glad you're liking it, @bproffitt! That gives me confidence. Some answers for you:

> Were there any conversion-related issues/cleanup concerns?

Conversion wasn't perfect, but involved pretty much the level of effort I expected it to. In the end, I had to touch each file manually to fix fiddly things, and you'll likely see some artifacts from that (I'm sure I didn't catch everything in a single pass). The issues that stand out to me are:

* GitBook likes `_` characters for designating italics in Markdown. When publishing through the GUI, it will replace all `*` characters used for this purpose with `_` ([example commit](https://github.com/theopensourceway/guidebook/commit/44569e86d5b6639451fc4715aade281da0890200)). Not really serious; just something we should note.
* The conversion undid our one-line-per-sentence formatting convention for source files. Each paragraph is now a line. I started correcting that at first, but then gave up—it would just have taken too long, and that's the kind of thing that will sort itself out in further editorial passes and revisions anyway. But I appreciate that convention we established and eventually want to restore it. I just hope GitBook does not consistently undo that work every time we push a new version from its GUI.
* Switching to GitBook may force us to confront our inconsistent use of hyperlinks versus endnotes. Some chapters use nothing but the former; some use nothing but the latter. Most use both. GitBook does not handle Markdown-formatted endnotes as well as I'd like it to. Instead, I had to use the GitBook GUI to manually re-add all endnotes back into each chapter after "typesetting" the plain text versions. Endnotes then get rendered as hover-over popups—pretty neat. But if we're going to continue using them (and if we're going to continue publishing alternate versions of the book, e.g. PDF and ePub, we should), then we need to work on refining our practice here.
* GitBook requires a different format for file frontmatter/metadata. It reads certain fields separated by `---` and `---` at the start of each file ([example commit](https://github.com/theopensourceway/guidebook/commit/5971b722d9fd8be4780bee1ad9bff918852770c2)). The `description` line gets reproduced as a chapter subtitle. You can also track `author` and `updated`. I had to learn these things by experimenting, because I don't see them documented anywhere! So I will keep looking for a complete list of available parameters we can define in the file frontmatter.

Those are the things that come to mind immediately. I'll follow up if anything else strikes me (likely to happen as soon as I step away).

> Do we care that the "Powered by Gitbook" stamp is all over the site? Part of me is like "cool, support our fellow open source projects" and part of me is also "it feels like a NASCAR sticker."

I agree that it's perhaps a little more intrusive than it needs to be. However, if we intend to continue with GitBook as our publishing platform, I don't think we have much choice in the matter. Since we're now (as in: as of today) on the [GitBook community plan](https://docs.gitbook.com/account-management/plans/apply-for-the-non-profit-open-source-plan), the banner is standard issue. We'll need to consider it our reciprocation for the freebies we receive if we choose to move forward here.

**Comment by semioticrobotic** [comment]
I've opened PR #241 so people can grasp a little more concrete what I'm advocating here.

**Comment by semioticrobotic** [comment]
Happy 2025, everyone! What would help us move this discussion forward in the new year? I am currently outlining priority projects for the coming months and would very much like to continue contributing here.

**Comment by bproffitt** [comment]
I'm going to need a moment to catch my collective breath and figure out
what 1Q 2025 looks like. I should be more settled in a week's time.

BKP

**Reply by semioticrobotic** [reply]
Totally understandable! Thanks for checking in. 😄

**Comment by semioticrobotic** [comment]
It's been a few months, so (and, I realize, at the risk of becoming sort of a pest) I figured I would revisit this discussion for a gut-check from other maintainers. @bproffitt @shaunix @quaid @samus-aran: I'm wondering what ya'll think about [where we stand](https://github.com/theopensourceway/guidebook/pull/241). I'm still very keen to refresh and push this project ahead in 2025.

**Comment by semioticrobotic** [comment]
Hearing no objections, then, I'll assume a lazy consensus and prepare to make these changes in the next few days, unless someone objects. I will:

- Merge #241 
- Designate `https://theopensourceway.gitbook.io/guidebook` the new canonical source of the guidebook
- Launch a new, simpler [project website](https://github.com/theopensourceway/website) using GitHub Pages
- Get to work on [open issues](https://github.com/theopensourceway/guidebook/issues)!

The only thing I will not be able to do is redirect our project URL, `theopensourceway.org`, to the new landing page.

Always grateful for feedback or guidance.

**Comment by bproffitt** [comment]
These all sound great. I apologize for the lack of feedback. As I mentioned to The Boss the other day, I (wrongly) assumed you were going to move forward. I regret the misunderstanding.

**Reply by semioticrobotic** [reply]
No sweat, @bproffitt. I'm just glad for the affirmation that we're on the right track here. I will work on the merge and follow-up items in just a bit. Meanwhile, can I lean on you to help with the domain's DNS update, assuming we wish to point to the updated and refreshed website with `theopensourceway.org`? I think your infra team is managing that one.

**Reply by bproffitt** [reply]
Yeah, just ping me directly and let me know what you need and when.

**Reply by semioticrobotic** [reply]
Email off to you, @bproffitt! Thanks for your willingness to investigate.

**Comment by semioticrobotic** [comment]
As of earlier today, [we are live](https://guidebook.theopensourceway.org)! Many thanks to all for your willingness to try this new direction. I think we can close this discussion.
